File: frontend/src/utils/chartKey.ts
/** Match backend `normalize_chart_key` for linking PDF charts to RNP procedures. */

export function normalizeChartKey(text: string | null | undefined): string {
  if (!text) return '';

  let s = text.trim();
  if (!s) return '';

  try {
    const u = new URL(s);
    if (u.protocol === 'http:' || u.protocol === 'https:') {
      const seg = u.pathname.split('/').filter(Boolean);
      s = seg.length ? decodeURIComponent(seg[seg.length - 1] ?? '') || s : s;
    }
  } catch {
    // not a full URL; keep s
  }

  if (s.includes('/')) {
    s = s.split('/').pop() ?? s;
  }

  s = s.toUpperCase();
  if (s.endsWith('.PDF')) s = s.slice(0, -4);

  s = s.replace(/[_\s]+/g, '-');
  s = s.replace(/-+/g, '-');

  const suffixes = [
    '-CODING',
    '-TABLES',
    '-TABLE',
    '-CAT-A-B-C-D',
    '-CAT-A-B-C',
    '-CAT-A-B',
    '-FAS-DATA',
    '-PROFILE',
  ];
  for (const suffix of suffixes) {
    if (s.includes(suffix)) {
      s = s.split(suffix).join('');
    }
  }

  s = s.replace(/-+/g, '-');
  s = s.replace(/^-+|-+$/g, '');

  return s;
}

File: backend/app/utils/chart_key.py
"""Normalize chart / procedure names for cross-linking PDF charts to DB rows."""

import re
from urllib.parse import unquote, urlparse


def normalize_chart_key(text: str | None) -> str:
    """
    Uppercase, strip trailing .pdf, collapse separators (-, _, whitespace) to a
    single '-', and remove other noise so chart filenames match procedure names.
    """
    if not text:
        return ""

    s = text.strip()
    if not s:
        return ""

    parsed = urlparse(s)
    if parsed.scheme in ("http", "https") and parsed.path:
        s = unquote(parsed.path.rsplit("/", 1)[-1])
    elif "/" in s:
        s = s.rsplit("/", 1)[-1]

    s = s.upper()
    if s.endswith(".PDF"):
        s = s[:-4]

    s = re.sub(r"[_\s]+", "-", s)
    s = re.sub(r"-+", "-", s)

    # Strip common secondary chart suffixes so tables/coding charts match the base procedure
    for suffix in [
        "-CODING",
        "-TABLES",
        "-TABLE",
        "-CAT-A-B-C-D",
        "-CAT-A-B-C",
        "-CAT-A-B",
        "-FAS-DATA",
        "-PROFILE",
    ]:
        if suffix in s:
            s = s.replace(suffix, "")

    s = re.sub(r"-+", "-", s)
    s = s.strip("-")

    return s
